File: python-sdk/agentscope/trace.py
from __future__ import annotations

from typing import Any, Dict

from .instrumentation import auto_trace
from .run import _append_artifact, _current_parent_span_id, _update_span


class _TraceFacade:
    def auto(self, providers: list[str] | None = None) -> None:
        auto_trace(providers=providers)

    def log(
        self,
        message: str,
        *,
        level: str = "info",
        span_id: str | None = None,
        metadata: Dict[str, Any] | None = None,
        timestamp: str | None = None,
    ) -> Dict[str, Any]:
        resolved_span_id = span_id or _current_parent_span_id()
        payload: Dict[str, Any] = {
            "message": message,
            "level": level,
        }
        if metadata is not None:
            payload["metadata"] = metadata
        if timestamp is not None:
            payload["timestamp"] = timestamp

        return _append_artifact("log", payload, span_id=resolved_span_id)

    def update_span(self, span_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        return _update_span(span_id, data)


trace = _TraceFacade()

File: ts-sdk/src/trace.ts
import { currentSpanId } from "./context";
import { autoTrace } from "./instrumentation";
import { addArtifact, updateSpan } from "./span";
import type { SpanRecord, TraceLogOptions } from "./types";

class TraceFacade {
  auto(providers?: string[]): void {
    autoTrace(providers);
  }

  log(message: string, options: TraceLogOptions = {}) {
    const payload: Record<string, unknown> = {
      message,
      level: options.level ?? "info",
    };

    if (options.metadata !== undefined) {
      payload.metadata = options.metadata;
    }
    if (options.timestamp !== undefined) {
      payload.timestamp = options.timestamp;
    }

    const resolvedSpanId = options.spanId === undefined ? currentSpanId() : options.spanId;
    return addArtifact("log", payload, resolvedSpanId);
  }

  updateSpan(spanId: string, data: Partial<SpanRecord>): SpanRecord {
    return updateSpan(spanId, data);
  }
}

export const trace = new TraceFacade();
